processor/metricstarttimeprocessor/internal/subtractinitial/adjuster.go

```diff
@@ -19,19 +22,354 @@ import (
+		adjustedPoint := pmetric.NewExponentialHistogramDataPoint()
+		currentDist.CopyTo(adjustedPoint)
+		adjustedPoint.SetStartTimestamp(referenceTsi.ExponentialHistogram.StartTimestamp())
+		if currentDist.Flags().NoRecordedValue() {
+			// TODO: Investigate why this does not reset.
+			tmp := resExpHistogram.DataPoints().AppendEmpty()
+			adjustedPoint.CopyTo(tmp)
+			continue
+		}
+		subtractExponentialHistogramDataPoint(adjustedPoint, referenceTsi.ExponentialHistogram)
+
+		previousTsi, found := previousValueTsm.Get(current, currentDist.Attributes())
+		if !found {
+			// First point after the reference. Not a reset.
+			previousTsi.ExponentialHistogram = pmetric.NewExponentialHistogramDataPoint()
+		} else if previousTsi.IsResetExponentialHistogram(adjustedPoint) {
+			// reset re-initialize everything and use the non adjusted points start time.
+			referenceTsi.ExponentialHistogram = pmetric.NewExponentialHistogramDataPoint()
+			referenceTsi.ExponentialHistogram.SetStartTimestamp(currentDist.StartTimestamp())
+
+			tmp := resExpHistogram.DataPoints().AppendEmpty()
+			currentDist.CopyTo(tmp)
+			continue
+		}
+
+		// Update previous values with the current point.
+		adjustedPoint.CopyTo(previousTsi.ExponentialHistogram)
+		tmp := resExpHistogram.DataPoints().AppendEmpty()
+		adjustedPoint.CopyTo(tmp)
+	}
+}
+
+func adjustMetricSum(referenceTsm, previousValueTsm *datapointstorage.TimeseriesMap, current pmetric.Metric, resSum pmetric.Sum) {
+	resSum.SetAggregationTemporality(current.Sum().AggregationTemporality())
+	resSum.SetIsMonotonic(current.Sum().IsMonotonic())
+
+	currentPoints := current.Sum().DataPoints()
+	for i := 0; i < currentPoints.Len(); i++ {
+		currentDist := currentPoints.At(i)
+
+		referenceTsi, found := referenceTsm.Get(current, currentDist.Attributes())
+		if !found {
+			// initialize everything. Don't add the datapoint to the result.
+			referenceTsi.Number = pmetric.NewNumberDataPoint()
+			currentDist.CopyTo(referenceTsi.Number)
+			continue
+		}
+
+		// Adjust the datapoint based on the reference value.
+		adjustedPoint := pmetric.NewNumberDataPoint()
+		currentDist.CopyTo(adjustedPoint)
+		adjustedPoint.SetStartTimestamp(referenceTsi.Number.StartTimestamp())
+		if currentDist.Flags().NoRecordedValue() {
+			// TODO: Investigate why this does not reset.
+			tmp := resSum.DataPoints().AppendEmpty()
+			adjustedPoint.CopyTo(tmp)
+			continue
+		}
+		adjustedPoint.SetDoubleValue(adjustedPoint.DoubleValue() - referenceTsi.Number.DoubleValue())
+
+		previousTsi, found := previousValueTsm.Get(current, currentDist.Attributes())
+		if !found {
+			// First point after the reference. Not a reset.
+			previousTsi.Number = pmetric.NewNumberDataPoint()
+		} else if previousTsi.IsResetSum(adjustedPoint) {
+			// reset re-initialize everything and use the non adjusted points start time.
+			referenceTsi.Number = pmetric.NewNumberDataPoint()
+			referenceTsi.Number.SetStartTimestamp(currentDist.StartTimestamp())
```

Comment:
Ah I think this is the root cause of some unexpected missing sample on https://github.com/bwplotka/otel-ops-col/pull/1. So generally this is an accurate flow, just we miss the opportunity to make experience more accurate and NOT skipping sample here.

See [the GMP code for the same](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/main/pkg/export/series_cache.go#L287), we have opportunity here to "de-normalize" the stream, so literally use the `currentDist` sample as the CORRECT sample value and start time being any timestamp between last sample and this sample (-1ms is likely the most accurate guess). 

This is what proven to work for thousands of GCM PromQL users, so I would vote for adding this optimization in, but up to you!

Reply:
Changed things to make sure the reset point has a -1ms start time.